Managers/UInputMan.cpp

```diff
@@ -1,2858 +1,1393 @@
+			mouseY = std::min(mouseY, y + height * g_FrameMan.ResolutionMultiplier());
 
-//////////////////////////////////////////////////////////////////////////////////////////
-// Method:          MenuButtonReleased
-//////////////////////////////////////////////////////////////////////////////////////////
-// Description:     Shows whether any generic button with the menu cursor is held down.
+			position_mouse(mouseX, mouseY);
+		}
+	}
 
-bool UInputMan::MenuButtonReleased(int whichButton)
-{
-    bool button = false;
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
-    // Go through all players
-    for (int player = 0; player < MAX_PLAYERS && !button; ++player)
-    {
-        // Which Device is used by this player
-        int device = m_aControlScheme[player].GetDevice();
+	void UInputMan::ForceMouseWithinPlayerScreen(short whichPlayer) {
+		if (whichPlayer < PLAYER_ONE || whichPlayer >= PLAYER_FOUR) {
+			return;
+		}
+		int screenWidth = g_FrameMan.GetPlayerFrameBufferWidth(whichPlayer);
+		int screenHeight = g_FrameMan.GetPlayerFrameBufferHeight(whichPlayer);
+
+		// If we are dealing with split screens, then draw the intermediate draw splitscreen to the appropriate spot on the back buffer
+		if (g_FrameMan.GetScreenCount() > 1) {
+			if (whichPlayer == 0) {
+				// Always upper left corner
+				ForceMouseWithinBox(0, 0, screenWidth, screenHeight, whichPlayer);
+			} else if (whichPlayer == 1) {
+				// If both splits, or just V-Split, then in upper right quadrant
+				if ((g_FrameMan.GetVSplit() && !g_FrameMan.GetHSplit()) || (g_FrameMan.GetVSplit() && g_FrameMan.GetHSplit())) {
+					ForceMouseWithinBox(g_FrameMan.GetResX() / 2, 0, screenWidth, screenHeight, whichPlayer);
+				} else {
+					ForceMouseWithinBox(0, g_FrameMan.GetResY() / 2, screenWidth, screenHeight, whichPlayer);
+				}
+			} else if (whichPlayer == 2) {
+				// Always lower left quadrant
+				ForceMouseWithinBox(0, g_FrameMan.GetResY() / 2, screenWidth, screenHeight, whichPlayer);
 
-        // Check for primary/secondary button presses
-		if (whichButton >= MENU_PRIMARY)
-			button = ElementReleased(player, INPUT_FIRE) || MouseButtonReleased(MOUSE_LEFT, player) || button;
-        if (whichButton >= MENU_SECONDARY)
-            button = ElementReleased(player, INPUT_PIEMENU) || MouseButtonReleased(MOUSE_RIGHT, player) || button;
-    }
+			} else if (whichPlayer == 3) {
+				// Always lower right quadrant
+				ForceMouseWithinBox(g_FrameMan.GetResX() / 2, g_FrameMan.GetResY() / 2, screenWidth, screenHeight, whichPlayer);
+			}
+		} else {
+			ForceMouseWithinBox(0, 0, screenWidth, screenHeight, whichPlayer);
+		}
+	}
 
-    return button;
-}
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
+	bool UInputMan::JoystickActive(int joystickNumber) const {
+		if (joystickNumber < 0 || joystickNumber >= MAX_PLAYERS) {
```

Comment:
num_joysticks is the number of currently active joysticks, which ranges from 1 to 8, with 0 being none active.
In this case num_joysticks acts as MaxJoystickCount.

This is kind of a weird check I guess because it's not tied to player numbers.
So for example if `Player 3` is the only one using a controller, and you check for `joystickNumber == 2`, this will return false because `num_joysticks == 1`, so you need to run a loop over players and do quick maffs to get the corresponding input device.
If this even makes any sense to you, cause i'm not even sure if I got it right.
You can see how it's used in MainMenuGUI.cpp line 1742.

All in all I feel like there needs to be a big bren full rewrite on joystick handling, probably a full on wrapper.
Probably something along the lines of making a joystick object and tie the index to a player number.
Not sure tho, this is just a 3 second thought.

Reply:
Ahhh okay, thanks for clarifying it. What you said makes sense, and looking briefly at the line in MainMenuGUI and stuff I think you've got it right.
I stand by onelining it and moving it to the header though, I think `return joystickNumber >= 0 && joystickNumber < MAX_PLAYERS && joystickNumber < num_joysticks;` is sufficiently short and readable.

A cleaner wrapper for joystick stuff would probably be nice someday, it's certainly weird, somewhat stuff. I guess UInputMan is already supposed to be that wrapper though haha.